Create API to await multiple PlayerLoopTiming without  any allocation

RamType-0 opened the pull request:
Now almost of parameters for synchronization in UniTask is represented by `SyncParams`.

We could await multiple timings by `UniTask.SwitchToSyncParams(SyncParams,CancellationToken)`.

He said
https://github.com/Cysharp/UniTask/issues/153#issuecomment-683652258

So I defined `SyncParams.MainThread = EarlyUpdate | PreUpdate | Update | PreLateUpdate | PostLateUpdate`.

And we could also switch to thread pool by `SyncParams.ThreadPool`.

While I implement it, I need functionality to know what `PlayerLoopTiming` we are currently in.
I also added it ,`PlayerLoopHelper.TryGetCurrentPlayerLoopTiming()`.
This returns null when we call it from non main thread.
This is implemented by combination of `ThreadStatic` and assigning current `PlayerLoopTiming` in `ContinuationQueue.RunCore()` and `PlayerLoopRunner.RunCore()`.

neuecc commented:
The implementation is great, but you're messing with PlayerLoop too much for limited use.
If there's no other use for it, I don't want to merge.

RamType-0 commented:
I didn't implemented many public API at here,but most of these are implemented through new low level(?) API `PlayerLoopHelper.AddAction(SyncParams,IPlayerLoopItem)` `PlayerLoopHelper.AddContinuation(SyncParams,Action)`, and `PlayerLoopHelper.TryGetCurrentPlayerLoopTiming()`.
This means we could also add multiple timing support for other method very easily(e.g. `UniTask.Yield(SyncParams)`).

neuecc commented:
In other words, there is no particular beneficial usage.

RamType-0 commented:
> He said
> [#153 (comment)](https://github.com/Cysharp/UniTask/issues/153#issuecomment-683652258)
> 
> So I defined `SyncParams.MainThread = EarlyUpdate | PreUpdate | Update | PreLateUpdate | PostLateUpdate`.

[But you also said LastPostLateUpdate might be a good addition.](https://github.com/Cysharp/UniTask/issues/127#issuecomment-673322248)

We have no definitely only one greatest timings for switching to main thread.
and this could be variant for how each application use `PlayerLoopTiming`s.

Currently,unity built-in async operation integrations are using inconsistent PlayerLoopTiming.
(Force using `PlayerLoopTiming.Update`,or using specified single `PlayerLoopTiming`,or using predefined multiple `PlayerLoopTiming` with allocation)
This feature will solve it.
If you feel this opinion is inconsistent between existence of `SyncParams.MainThread`, I will remove `SyncParams.MainThread`.

Also,`PlayerLoopHelper.TryGetCurrentPlayerLoopTiming()` is definitely required for `UniTask.SwitchToMainThread(PlayerLoopTiming)` which I proposed in #151 .

If only `PlayerLoopHelper.TryGetCurrentPlayerLoopTiming()` was enabled,I can implement
`UniTask.SwitchToSyncParams(SyncParams)` completely independently.

neuecc commented:
I think that PlayerLoop's "Somewhere" is less practical.
I keep hearing the same talk.
There doesn't seem to be any useful examples.

RamType-0 commented:
> I think that PlayerLoop's "Somewhere" is less practical.

Really? So why did you have PlayerLoopTiming in UniTask, and also it is extended from UniTask v1?
And why did you have additional effort to know current PlayerLoopTiming in ContinuationQueue and PlayerLoopRunner ? (It injects method with PlayerLoopTiming name into stack trace in debug build)
Cannot know current PlayerLoopTiming means we cannot know what kind of event(e.g. physics update,rendering) happens while we awaiting UniTask.Yield(PlayerLoopTiming).
This could cause one frame of tearing(e.g. Unexpected T pose shown on first one frame after asynchronously loaded model,one frame of chance to talking to NPC with some other window opened)


Checking current PlayerLoopTiming manually is seems to be bother,but very important for some Awaitable's internal implementation.

I added many APIs to PlayerLoopHelper,and you said "you're messing with PlayerLoop too much",
but what I really need to inject to existing UniTask system is only `PlayerLoopHelper.TryGetCurrentPlayerLoopTiming()`(and set current PlayerLoopTiming in ContinuationQueue.RunCore() and PlayerLoopRunner.RunCore()).
Most of other feature doesn't need any injection.

I just added many API into existing class because I didn't think that you don't want to merge "the great implementation" just because it is "messing with PlayerLoop too much".

RamType-0 commented:
JobHandle's await causes allocation every times,although jobs are used for performance critical scenario.
And this solve it.
Isn't it enough?

neuecc commented:
I think you're fundamentally mistaken,
the ability to wait for a frame somewhere is unnecessary.
JobHandle is such an implementation, but that's the special case.
And I don't think it's a good design.
However, I don't think it's that harmful, so I'm leaving it as is, but if it really bothers you, I'll remove it.

RamType-0 commented:
I'm really not sure about why is this feature is unnecessary ,although UniTask.Yield and UniTask.NextFrame were both needed.
And how is it messing PlayerLoop with too much?
This feature doesn't change any existing API.

UniTaskPlayerLoopSubSystem is just an IPlayerLoopItem which register for multiple timing,and do same thing as ContinuationQueue and PlayerLoopRunner when MoveNext is called.

If you think that any many addition of code is messing, I can reduce it to just 20 lines.Which only injects PlayerLoopHelper.TryGetCurrentPlayerLoopTiming().

RamType-0 commented:
I just faced a new issue which requires to await multiple PlayerLoopTiming.
I'm using external device which sends encoded data.
When I await to receive it,it could be done at thread pool.
After I receive it,I want to decode it as soon as possible using a job.
Because we could not schedule a job from thread pool,I need to return to main thread as soon as possible.

neuecc commented:
It's not same worth as Yield or NextFrame.
Obviously, it's not as beneficial........

Well, TryGetCurrentPlayerLoopTiming is nice to have.
I was going to include it.

However, I think it should be Get instead of Try (no need to stop getting it just because you're on a different thread).
Also, with that, `UniTask.NextPlayerLoop` could be implemented.
However, it is not go to Next correctly, depending on the timing when you're in a different thread.
It may need more implement to prevent that(Or don't care.).
